toolchain/check/modifiers_allowed.cpp

```diff
@@ -0,0 +1,47 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/modifiers_allowed.h"
+
+namespace Carbon::Check {
+
+static auto ReportNotAllowed(Context& context, Parse::Node modifier_node,
+                             llvm::StringRef decl_name,
+                             Parse::Node context_node) {
+  CARBON_DIAGNOSTIC(ModifierNotAllowedOn, Error, "`{0}` not allowed on {1}.",
+                    llvm::StringRef, llvm::StringRef);
+  if (context_node == Parse::Node::Invalid) {
+    context.emitter().Emit(modifier_node, ModifierNotAllowedOn,
+                           context.TextForNode(modifier_node), decl_name);
+  } else {
+    CARBON_DIAGNOSTIC(ModifierNotInContext, Note,
+                      "Containing definition here.");
+    context.emitter()
+        .Build(modifier_node, ModifierNotAllowedOn,
+               context.TextForNode(modifier_node), decl_name)
+        .Note(context_node, ModifierNotInContext)
+        .Emit();
+  }
```

Comment:
```suggestion
  auto diag = context.emitter()
        .Build(modifier_node, ModifierNotAllowedOn,
               context.TextForNode(modifier_node), decl_name);
  if (context_node.is_valid()) {
    CARBON_DIAGNOSTIC(ModifierNotInContext, Note,
                      "Containing definition here.");
    diag.Note(context_node, ModifierNotInContext)
  }
  diag.Emit();
```

This kind of approach allows less duplication of the diagnostic construction.

Also note the `is_valid()` check instead of `== Parse::Node::Invalid`.